Added adding and deleting todos

vova-sherstov opened the pull request:
[DEMO LINK}(https://vova-sherstov.github.io/react_todo-app-add-and-delete/)

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,24 +1,150 @@
+  const [errorMessage, setErrorMessage] = useState(ErrorMessages.None);
+
+  const preparedTodos = useMemo(() => {
+    return prepareTodos(todos, filterOption);
```
Consider renaming this function to show that it's an action
```suggestion
    return getPreparedTodos(todos, filterOption);
```

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,24 +1,150 @@
+    return prepareTodos(todos, filterOption);
+  }, [todos, filterOption]);
+
+  const allTodoCompleted = todos.every(todo => todo.completed);
```
In my opinion, it's better to check if there is at least one active todos and use some() method for better performance

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,24 +1,150 @@
+  };
+
+  useEffect(() => {
+    deleteTodo();
+  }, [deletingTodoIds]);
```
Would be better to wrap the deleteTodo to useCallback() to memorize it. And change the dependencies for this useEffect. And you really need to do it with useEffect? 🤔  Seems that deleting should start on the button click 
```suggestion
    deleteTodo();
  }, [deleteTodo]);
```

Anastasiia-Svintsova commented on src/App.tsx:
```diff
@@ -1,24 +1,150 @@
+  }, [deletingTodoIds]);
+
+  useEffect(() => {
+    setTimeout(() => {
```
You should clear all timers on the component unmount to avoid unpredictable side effects https://bobbyhadz.com/blog/react-cleartimeout

Anastasiia-Svintsova commented on src/components/TodoForm/TodoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+    if (hasTodo) {
+      titleInput.current?.focus();
+    }
+  }, [isDisabled]);
```
Don't you need to include `hasTodo` in dependencies?

vova-sherstov replied:
If i include hasTodo in dependencies the input field will be in focus immediately

TarasHoliuk replied:
In this case, there should be one more variable to decide if you want to focus it
because current behavior is a bit unexpected when you read code

It is not a critical thing I guess - try to fix it in the next part of the task

Anastasiia-Svintsova commented on src/components/TodoList/TodoList.tsx:
```diff
@@ -0,0 +1,101 @@
+                  type="text"
+                  className="todo__title-field"
+                  placeholder="Empty todo will be deleted"
+                  value={todo.title}
```
Use destructuring for objects for better code readability

Anastasiia-Svintsova reviewed (changes requested):
Keep going! 

- The user shouldn't be able to add a todo with an empty title 

https://github.com/mate-academy/react_todo-app-add-and-delete/assets/99403198/2c9f0fd7-1975-493b-8d8c-ba0a27786209

vova-sherstov commented:
Thank you for your feedback!
1. fixed adding a todo with an empty title;
2. renamed function;
3. allTodoCompleted replaced by hasActiveTodo using the some() method;
4. changed the todo delete method (I used useEffect hook because the deletion function passed in props did not work by ;clicking. Loader covered todo, but they were not deleted);
5. added clearTimeout() method and moved this useEffect hook to ErrorMessage component;

TarasHoliuk commented on src/components/TodoList/TodoList.tsx:
```diff
@@ -0,0 +1,100 @@
+      })}
+
+      {tempTodoTitle && (
+        <div className="todo temp-item-enter temp-item-enter-active">
```
to avoid duplication and to separate more complex logic in the next task don't forget to move it to a separate component

TarasHoliuk reviewed (approved):
GJ :+1: 
Added a few minor comments
just take them into account in the next task
